File: src/data_formats/parsing/itempmt.ts
import { BufferCursor } from "../BufferCursor";

export type ItemPmt = {
    stat_boosts: PmtStatBoost[];
    armors: PmtArmor[];
    shields: PmtShield[];
    units: PmtUnit[];
    tools: PmtTool[][];
    weapons: PmtWeapon[][];
};

export type PmtStatBoost = {
    stat_1: number;
    stat_2: number;
    amount_1: number;
    amount_2: number;
};

export type PmtWeapon = {
    id: number;
    type: number;
    skin: number;
    team_points: number;
    class: number;
    reserved_1: number;
    min_atp: number;
    max_atp: number;
    req_atp: number;
    req_mst: number;
    req_ata: number;
    mst: number;
    max_grind: number;
    photon: number;
    special: number;
    ata: number;
    stat_boost: number;
    projectile: number;
    photon_trail_1_x: number;
    photon_trail_1_y: number;
    photon_trail_2_x: number;
    photon_trail_2_y: number;
    photon_type: number;
    unknown_1: number[];
    tech_boost: number;
    combo_type: number;
};

export type PmtArmor = {
    id: number;
    type: number;
    skin: number;
    team_points: number;
    dfp: number;
    evp: number;
    block_particle: number;
    block_effect: number;
    class: number;
    reserved_1: number;
    required_level: number;
    efr: number;
    eth: number;
    eic: number;
    edk: number;
    elt: number;
    dfp_range: number;
    evp_range: number;
    stat_boost: number;
    tech_boost: number;
    unknown_1: number;
};

export type PmtShield = PmtArmor;

export type PmtUnit = {
    id: number;
    type: number;
    skin: number;
    team_points: number;
    stat: number;
    stat_amount: number;
    plus_minus: number;
    reserved: number[];
};

export type PmtTool = {
    id: number;
    type: number;
    skin: number;
    team_points: number;
    amount: number;
    tech: number;
    cost: number;
    item_flag: number;
    reserved: number[];
};

export function parse_item_pmt(cursor: BufferCursor): ItemPmt {
    cursor.seek_end(32);
    const main_table_offset = cursor.u32();
    const main_table_size = cursor.u32();
    // const main_table_count = cursor.u32(); // Should be 1.

    cursor.seek_start(main_table_offset);

    const compact_table_offsets = cursor.u16_array(main_table_size);
    const table_offsets: { offset: number; size: number }[] = [];
    let expanded_offset = 0;

    for (const compact_offset of compact_table_offsets) {
        expanded_offset = expanded_offset + 4 * compact_offset;
        cursor.seek_start(expanded_offset - 4);
        const size = cursor.u32();
        const offset = cursor.u32();
        table_offsets.push({ offset, size });
    }

    const item_pmt: ItemPmt = {
        // This size (65268) of this table seems wrong, so we pass in a hard-coded value.
        stat_boosts: parse_stat_boosts(cursor, table_offsets[305].offset, 52),
        armors: parse_armors(cursor, table_offsets[7].offset, table_offsets[7].size),
        shields: parse_shields(cursor, table_offsets[8].offset, table_offsets[8].size),
        units: parse_units(cursor, table_offsets[9].offset, table_offsets[9].size),
        tools: [],
        weapons: [],
    };

    for (let i = 11; i <= 37; i++) {
        item_pmt.tools.push(parse_tools(cursor, table_offsets[i].offset, table_offsets[i].size));
    }

    for (let i = 38; i <= 275; i++) {
        item_pmt.weapons.push(
            parse_weapons(cursor, table_offsets[i].offset, table_offsets[i].size)
        );
    }

    return item_pmt;
}

function parse_stat_boosts(cursor: BufferCursor, offset: number, size: number): PmtStatBoost[] {
    cursor.seek_start(offset);
    const stat_boosts: PmtStatBoost[] = [];

    for (let i = 0; i < size; i++) {
        stat_boosts.push({
            stat_1: cursor.u8(),
            stat_2: cursor.u8(),
            amount_1: cursor.i16(),
            amount_2: cursor.i16(),
        });
    }

    return stat_boosts;
}

function parse_weapons(cursor: BufferCursor, offset: number, size: number): PmtWeapon[] {
    cursor.seek_start(offset);
    const weapons: PmtWeapon[] = [];

    for (let i = 0; i < size; i++) {
        weapons.push({
            id: cursor.u32(),
            type: cursor.i16(),
            skin: cursor.i16(),
            team_points: cursor.i32(),
            class: cursor.u8(),
            reserved_1: cursor.u8(),
            min_atp: cursor.i16(),
            max_atp: cursor.i16(),
            req_atp: cursor.i16(),
            req_mst: cursor.i16(),
            req_ata: cursor.i16(),
            mst: cursor.i16(),
            max_grind: cursor.u8(),
            photon: cursor.i8(),
            special: cursor.u8(),
            ata: cursor.u8(),
            stat_boost: cursor.u8(),
            projectile: cursor.u8(),
            photon_trail_1_x: cursor.i8(),
            photon_trail_1_y: cursor.i8(),
            photon_trail_2_x: cursor.i8(),
            photon_trail_2_y: cursor.i8(),
            photon_type: cursor.i8(),
            unknown_1: cursor.u8_array(5),
            tech_boost: cursor.u8(),
            combo_type: cursor.u8(),
        });
    }

    return weapons;
}

function parse_armors(cursor: BufferCursor, offset: number, size: number): PmtArmor[] {
    cursor.seek_start(offset);
    const armors: PmtArmor[] = [];

    for (let i = 0; i < size; i++) {
        armors.push({
            id: cursor.u32(),
            type: cursor.i16(),
            skin: cursor.i16(),
            team_points: cursor.i32(),
            dfp: cursor.i16(),
            evp: cursor.i16(),
            block_particle: cursor.u8(),
            block_effect: cursor.u8(),
            class: cursor.u8(),
            reserved_1: cursor.u8(),
            required_level: cursor.u8(),
            efr: cursor.u8(),
            eth: cursor.u8(),
            eic: cursor.u8(),
            edk: cursor.u8(),
            elt: cursor.u8(),
            dfp_range: cursor.u8(),
            evp_range: cursor.u8(),
            stat_boost: cursor.u8(),
            tech_boost: cursor.u8(),
            unknown_1: cursor.i16(),
        });
    }

    return armors;
}

function parse_shields(cursor: BufferCursor, offset: number, size: number): PmtShield[] {
    return parse_armors(cursor, offset, size);
}

function parse_units(cursor: BufferCursor, offset: number, size: number): PmtUnit[] {
    cursor.seek_start(offset);
    const units: PmtUnit[] = [];

    for (let i = 0; i < size; i++) {
        units.push({
            id: cursor.u32(),
            type: cursor.i16(),
            skin: cursor.i16(),
            team_points: cursor.i32(),
            stat: cursor.i16(),
            stat_amount: cursor.i16(),
            plus_minus: cursor.u8(),
            reserved: cursor.u8_array(3),
        });
    }

    return units;
}

function parse_tools(cursor: BufferCursor, offset: number, size: number): PmtTool[] {
    cursor.seek_start(offset);
    const tools: PmtTool[] = [];

    for (let i = 0; i < size; i++) {
        tools.push({
            id: cursor.u32(),
            type: cursor.i16(),
            skin: cursor.i16(),
            team_points: cursor.i32(),
            amount: cursor.i16(),
            tech: cursor.i16(),
            cost: cursor.i32(),
            item_flag: cursor.u8(),
            reserved: cursor.u8_array(3),
        });
    }

    return tools;
}

File: psolib/src/commonMain/kotlin/world/phantasmal/psolib/fileFormats/ItemPmt.kt
package world.phantasmal.psolib.fileFormats

import world.phantasmal.psolib.cursor.Cursor

class ItemPmt(
    val statBoosts: List<PmtStatBoost>,
    val frames: List<PmtFrame>,
    val barriers: List<PmtBarrier>,
    val units: List<PmtUnit>,
    val tools: List<List<PmtTool>>,
    val weapons: List<List<PmtWeapon>>,
)

class PmtStatBoost(
    val stat1: Int,
    val stat2: Int,
    val amount1: Int,
    val amount2: Int,
)

class PmtFrame(
    val id: Int,
    val type: Int,
    val skin: Int,
    val teamPoints: Int,
    val dfp: Int,
    val evp: Int,
    val blockParticle: Int,
    val blockEffect: Int,
    val frameClass: Int,
    val reserved1: Int,
    val requiredLevel: Int,
    val efr: Int,
    val eth: Int,
    val eic: Int,
    val edk: Int,
    val elt: Int,
    val dfpRange: Int,
    val evpRange: Int,
    val statBoost: Int,
    val techBoost: Int,
    val unknown1: Int,
)

typealias PmtBarrier = PmtFrame

class PmtUnit(
    val id: Int,
    val type: Int,
    val skin: Int,
    val teamPoints: Int,
    val stat: Int,
    val statAmount: Int,
    val plusMinus: Int,
    val reserved: ByteArray,
)

class PmtTool(
    val id: Int,
    val type: Int,
    val skin: Int,
    val teamPoints: Int,
    val amount: Int,
    val tech: Int,
    val cost: Int,
    val itemFlag: Int,
    val reserved: ByteArray,
)

class PmtWeapon(
    val id: Int,
    val type: Int,
    val skin: Int,
    val teamPoints: Int,
    val weaponClass: Int,
    val reserved1: Int,
    val minAtp: Int,
    val maxAtp: Int,
    val reqAtp: Int,
    val reqMst: Int,
    val reqAta: Int,
    val mst: Int,
    val maxGrind: Int,
    val photon: Int,
    val special: Int,
    val ata: Int,
    val statBoost: Int,
    val projectile: Int,
    val photonTrail1x: Int,
    val photonTrail1y: Int,
    val photonTrail2x: Int,
    val photonTrail2y: Int,
    val photonType: Int,
    val unknown1: ByteArray,
    val techBoost: Int,
    val comboType: Int,
)

fun parseItemPmt(cursor: Cursor): ItemPmt {
    val index = parseRel(cursor, parseIndex = true).index

    // This size (65268) of this table seems wrong, so we pass in a hard-coded value.
    val statBoosts = parseStatBoosts(cursor, index[305].offset, 52)
    val frames = parseFrames(cursor, index[7].offset, index[7].size)
    val barriers = parseFrames(cursor, index[8].offset, index[8].size)
    val units = parseUnits(cursor, index[9].offset, index[9].size)
    val tools = mutableListOf<List<PmtTool>>()
    val weapons = mutableListOf<List<PmtWeapon>>()

    for (i in 11..37) {
        tools.add(parseTools(cursor, index[i].offset, index[i].size))
    }

    for (i in 38..275) {
        weapons.add(parseWeapons(cursor, index[i].offset, index[i].size))
    }

    return ItemPmt(
        statBoosts,
        frames,
        barriers,
        units,
        tools,
        weapons,
    )
}

private fun parseStatBoosts(cursor: Cursor, offset: Int, size: Int): List<PmtStatBoost> {
    cursor.seekStart(offset)

    val statBoosts = mutableListOf<PmtStatBoost>()

    repeat(size) {
        statBoosts.add(PmtStatBoost(
            stat1 = cursor.uByte().toInt(),
            stat2 = cursor.uByte().toInt(),
            amount1 = cursor.short().toInt(),
            amount2 = cursor.short().toInt(),
        ))
    }

    return statBoosts
}

private fun parseFrames(cursor: Cursor, offset: Int, size: Int): List<PmtFrame> {
    cursor.seekStart(offset)

    val frames = mutableListOf<PmtFrame>()

    repeat(size) {
        frames.add(PmtFrame(
            id = cursor.int(),
            type = cursor.short().toInt(),
            skin = cursor.short().toInt(),
            teamPoints = cursor.int(),
            dfp = cursor.short().toInt(),
            evp = cursor.short().toInt(),
            blockParticle = cursor.uByte().toInt(),
            blockEffect = cursor.uByte().toInt(),
            frameClass = cursor.uByte().toInt(),
            reserved1 = cursor.uByte().toInt(),
            requiredLevel = cursor.uByte().toInt(),
            efr = cursor.uByte().toInt(),
            eth = cursor.uByte().toInt(),
            eic = cursor.uByte().toInt(),
            edk = cursor.uByte().toInt(),
            elt = cursor.uByte().toInt(),
            dfpRange = cursor.uByte().toInt(),
            evpRange = cursor.uByte().toInt(),
            statBoost = cursor.uByte().toInt(),
            techBoost = cursor.uByte().toInt(),
            unknown1 = cursor.short().toInt(),
        ))
    }

    return frames
}

private fun parseUnits(cursor: Cursor, offset: Int, size: Int): List<PmtUnit> {
    cursor.seekStart(offset)

    val units = mutableListOf<PmtUnit>()

    repeat(size) {
        units.add(PmtUnit(
            id = cursor.int(),
            type = cursor.short().toInt(),
            skin = cursor.short().toInt(),
            teamPoints = cursor.int(),
            stat = cursor.short().toInt(),
            statAmount = cursor.short().toInt(),
            plusMinus = cursor.uByte().toInt(),
            reserved = cursor.byteArray(3),
        ))
    }

    return units
}

private fun parseTools(cursor: Cursor, offset: Int, size: Int): List<PmtTool> {
    cursor.seekStart(offset)

    val tools = mutableListOf<PmtTool>()

    repeat(size) {
        tools.add(PmtTool(
            id = cursor.int(),
            type = cursor.short().toInt(),
            skin = cursor.short().toInt(),
            teamPoints = cursor.int(),
            amount = cursor.short().toInt(),
            tech = cursor.short().toInt(),
            cost = cursor.int(),
            itemFlag = cursor.uByte().toInt(),
            reserved = cursor.byteArray(3),
        ))
    }

    return tools
}

private fun parseWeapons(cursor: Cursor, offset: Int, size: Int): List<PmtWeapon> {
    cursor.seekStart(offset)

    val weapons = mutableListOf<PmtWeapon>()

    repeat(size) {
        weapons.add(PmtWeapon(
            id = cursor.int(),
            type = cursor.short().toInt(),
            skin = cursor.short().toInt(),
            teamPoints = cursor.int(),
            weaponClass = cursor.uByte().toInt(),
            reserved1 = cursor.uByte().toInt(),
            minAtp = cursor.short().toInt(),
            maxAtp = cursor.short().toInt(),
            reqAtp = cursor.short().toInt(),
            reqMst = cursor.short().toInt(),
            reqAta = cursor.short().toInt(),
            mst = cursor.short().toInt(),
            maxGrind = cursor.uByte().toInt(),
            photon = cursor.byte().toInt(),
            special = cursor.uByte().toInt(),
            ata = cursor.uByte().toInt(),
            statBoost = cursor.uByte().toInt(),
            projectile = cursor.uByte().toInt(),
            photonTrail1x = cursor.byte().toInt(),
            photonTrail1y = cursor.byte().toInt(),
            photonTrail2x = cursor.byte().toInt(),
            photonTrail2y = cursor.byte().toInt(),
            photonType = cursor.byte().toInt(),
            unknown1 = cursor.byteArray(5),
            techBoost = cursor.uByte().toInt(),
            comboType = cursor.uByte().toInt(),
        ))
    }

    return weapons
}
